### PR TITLE
wasmer2: do not re-validate wasm code after instrumentation

### PR DESCRIPTION
**Pull request by Ekleog-NEAR**
Performance characteristics: between 5% and 15% wall-clock gain depending on the test sample. See https://github.com/near/nearcore-private/issues/1 for more details

Safety characteristics: I ran a comparative fuzzer between wasmer2 with and without secondary validation for 2 weeks on an older commit, plus 4 days on this specific commit to validate the rebase. The fuzzer was not able to find any difference in behavior between the two.

In addition, as far as I can tell validation already happens in prepare_contract’s call to validate_contract (within wasmparser_decode at the validator.validate(&body) line), so removing this second validation should be fine.

However, it does mean that we accept including instrumentation within our "trusted code base" of parts that if they fail could cause harm. In theory wasmer2 should not cause UB if passed unvalidated input, but even if this is true it could definitely panic if passed unvalidated input.

As such, this is still a somewhat risky change to make. However, seeing how instrumentation is relatively simple wasm code addition and how it passed meticulous fuzzing, I think it has been derisked enough for the performance gains to outweigh the third-layer-of-security loss.

### CODE REVIEW

**Comment by Ekleog-NEAR** [comment]
Oh also: I think this change makes sense in the grand scheme of things where we’re currently considering hoisting instrumentation directly into assembly code generation rather than going through webassembly again: then, the secondary validation would really achieve nothing.

File: runtime/near-vm-runner/src/wasmer2_runner.rs [hunk 1]
```diff
@@ -276,9 +276,6 @@ impl Wasmer2VM {
         let prepared_code = prepare::prepare_contract(code.code(), &self.config)
             .map_err(CompilationError::PrepareError)?;
 
-        self.engine
-            .validate(&prepared_code)
-            .map_err(|e| CompilationError::WasmerCompileError { msg: e.to_string() })?;
```
Comment (matklad) [inline runtime/near-vm-runner/src/wasmer2_runner.rs:281]: Let's turn this into a debug assert explaining why we believe we don't need extra validation here?

**Comment by nagisa** [comment]
To summarize above: this is definitely a good idea, but there’s some groundwork necessary before this change can land.

**Comment by Ekleog-NEAR** [comment]
For your first point, we’re already using wasmparser 0.78 in both wasmer2 and nearcore, and I don’t think we can keep them more in sync until we hoist wasmer2 inside nearcore, at which point we could use `workspace.wasmparser = true`.

The second point is a great one, and we should definitely have a fuzzer runner that does this and can pass in invalid wasm! I’ll get to it once I finish fixing up the tests.

As for the CI failure, I think I found the cause but it’s weird: it seems to me like… our instrumentation adds a memory block? It seems weird to me, but `wasm2wat` tells me that the bytes that are being passed to `validate_contract` (in nearcore, with the bytes that are in the .wasm() of the `test_multiple_memories` test) have a single memory block defined:
```
$ wasm2wat /tmp/validate_contract.wasm
(module
  (import "env" "" (memory (;0;) 239 248))
  (table (;0;) 524431 funcref)
  (export "\00\11\11\11\11\11\11\02" (memory 0)))
```
while the ones that are passed to `validate_module` (so in wasmer2) have multiple memory blocks defined:
```
$ wasm2wat /tmp/validate_module.wasm
/tmp/validate_module.wasm:000002e: error: only one memory block allowed
```
and AFAICT the only difference between these two wasm blobs should be instrumentation

I’ll be back investigating this tomorrow, but if any of you reads this and has ideas for why this is happening, I’d love to hear them :)

**Comment by nagisa** [comment]
https://github.com/near/nearcore/issues/6042

**Comment by Ekleog-NEAR** [comment]
I’ve just implemented the fuzzer and handled the review comments; with #6042 now being fixed I think this is now ready to roll? :)

**Comment by Ekleog-NEAR** [comment]
@nagisa Ping for review? :)

**Review by nagisa — APPROVED** [review]

**Pull request merged**